File: js/src/utils/compression.ts
import { isBrowser } from './environment';

async function compressBrowser(data: string): Promise<string> {
  const encoder = new TextEncoder();
  const dataBuffer = encoder.encode(data);

  const compressionStream = new CompressionStream('gzip');
  const writer = compressionStream.writable.getWriter();
  writer.write(dataBuffer);
  writer.close();

  const compressedData = await new Response(compressionStream.readable).arrayBuffer();
  const uint8Array = new Uint8Array(compressedData);

  // Convertir a base64
  let binary = '';
  for (let i = 0; i < uint8Array.length; i++) {
    binary += String.fromCharCode(uint8Array[i]);
  }
  return btoa(binary);
}

async function compressNode(data: string): Promise<string> {
  const { gzipSync } = await import('zlib');
  const compressed = gzipSync(Buffer.from(data, 'utf-8'));
  return compressed.toString('base64');
}

export async function compressAndEncode(data: string): Promise<string> {
  if (isBrowser()) {
    return compressBrowser(data);
  }
  return compressNode(data);
}

File: go/utils/compression.go
package utils

import (
	"bytes"
	"compress/gzip"
	"encoding/base64"
)

func CompressAndEncode(data string) (string, error) {
	var buf bytes.Buffer
	
	gzipWriter := gzip.NewWriter(&buf)
	if _, err := gzipWriter.Write([]byte(data)); err != nil {
		return "", err
	}
	
	if err := gzipWriter.Close(); err != nil {
		return "", err
	}
	
	compressed := buf.Bytes()
	encoded := base64.StdEncoding.EncodeToString(compressed)
	
	return encoded, nil
}
